### PR TITLE
Add users , authentication and authorization

### PR DESCRIPTION
**Pull request by Michaeljogoh**
Add authentication and authorization

### CODE REVIEW

File: controllers/authController.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        res.status(422).json({error:"password does not match"})
+
+    }
+
+    if(password < 6){
+       res.status(422).json({error:"password must not be less than six characters"})
+    }
+    
+    await  Users.findOne({email : email})
+    if(email){
```
Comment (Asylcreek) [inline controllers/authController.js:26]: Where is this `email` coming from?

File: controllers/authController.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        res.status(422).json({error:"Email already registered"})
+     } else {
+        
+
+    //Hash password
+const  encryptPassword = await bcrypt.hash(password, 10);
+
+const newUser = new Users.create({name , email , password:encryptPassword});
```
Comment (Asylcreek) [inline controllers/authController.js:34]: `new Users.` or `await Users.`?

File: controllers/authController.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    }
+    
+    await  Users.findOne({email : email})
+    if(email){
+        res.status(422).json({error:"Email already registered"})
+     } else {
+        
+
+    //Hash password
+const  encryptPassword = await bcrypt.hash(password, 10);
```
Comment (Asylcreek) [inline controllers/authController.js:32]: `encryptPassword` sounds like a function that helps you encrypt a password. You can come up with a better name.

File: controllers/authController.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+const loginUsers = async  (req , res ) =>{
+    const {email , password} = req.body;
+
+    if(!email|| !password){
+       res.status(400).json("please add email or password")
+    }
+
+const savedUser =  await Users.findOne({email:email})
+        if(!savedUser){
+         res.status(404).json({error:"Invalid Username"})
```
Comment (Asylcreek) [inline controllers/authController.js:56]: `Invalid username` or `Invalid email`?

File: controllers/authController.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+const doMatch =  await  bcrypt.compare(password, savedUser.password)
+            if(doMatch){
+               const token = jwt.sign({user_id:savedUser}, process.env.JWT_SECRET , {expiresIn : '1d'})
```
Comment (Asylcreek) [inline controllers/authController.js:61]: You're still saving all the user's info in the JWT?

**Review by Asylcreek — CHANGES_REQUESTED** [review]
You're still using 422 status codes!!

And I do not think I have seen all your other files, like the UserModel, etc

File: controllers/authController.js (hunk 1 above)
Comment (Michaeljogoh) [reply controllers/authController.js:26]: The email is coming from the  schema , to validate if it already exist

File: controllers/authController.js (hunk 2 above)
Comment (Michaeljogoh) [reply controllers/authController.js:34]: await Users.create({ })

File: controllers/authController.js (hunk 3 above)
Comment (Michaeljogoh) [reply controllers/authController.js:32]: Ok , i will use hashedPassword

File: controllers/authController.js (hunk 4 above)
Comment (Michaeljogoh) [reply controllers/authController.js:56]: Invalid email

File: controllers/authController.js (hunk 5 above)
Comment (Michaeljogoh) [reply controllers/authController.js:61]: Yes sir

File: controllers/authController.js (hunk 1 above)
Comment (Asylcreek) [reply controllers/authController.js:26]: Should it not come from the result of line 25 above?

File: controllers/authController.js (hunk 2 above)
Comment (Asylcreek) [reply controllers/authController.js:34]: Do the needful

File: controllers/authController.js (hunk 5 above)
Comment (Asylcreek) [reply controllers/authController.js:61]: Why?

File: controllers/authController.js (hunk 1 above)
Comment (Michaeljogoh) [reply controllers/authController.js:26]: Ohh  , i will fix it

File: controllers/authController.js (hunk 5 above)
Comment (Michaeljogoh) [reply controllers/authController.js:61]: I am signing a jwt token to the  user by user_id

Comment (Asylcreek) [reply controllers/authController.js:61]: That's not what your code is showing.

What does `savedUser` look like? Is it the entire user object or just the user id?

If it is the entire user object, why are you 'signing' the entire user object into the JWT?

**Pull request merged**